### PR TITLE
feat: network: add nv17 and integrate the corresponding go state type

### PR DESCRIPTION
**Pull request by jennijuju**
## Related Issues

Towards #9246, #9248, #9249

## Proposed Changes

- Upgrades go-state-types to include v9 types and migration
- Adds network version 17 and the upgrade height
- Adds the v17 migration
- Updates to FVM release/v2 (through FFI)
  - This includes reviving https://github.com/filecoin-project/lotus/pull/9134/files, which is incorporated into this PR
- Adds the v9 actors bundles (using https://github.com/filecoin-project/builtin-actors/releases/tag/dev%2F20220907-v9 for now)
- Refactors the Actor Registry creation process
  - This is something we cut from scope in v16 by instead using the v8 specs-actors, and monkey-patching the synthetic -> real CIDs
  - With this PR, we get the method signatures from go-state-types (see https://github.com/filecoin-project/go-state-types/pull/66)
  - We use that information to create new `RegistryEntry`s that satisfy the same interface
  - There is a downgrade in UX experience here: we no longer have the method names in the registry
  - I don't know how to include this info, suggestions welcome. For now, we're including the MethodNum instead.
- Handles most of the integration work needed for FIP-0034
  - We use the new v9 miner types wherever possible
  - This mostly requires handling the fact that `SectorPreCommitInfo` and `PreCommitSectorParams` are no longer the same type
  - We simply create `PreCommitSectorParams` from `SectorPreCommitInfo` wherever necessary
  - We also need to recalibrate the calculation of necessary Pre-Commit Deposit when estimating collateral needed for miners
- Handles most of the integration work needed for FIP-0041 (by skipping it for now)
  - We do NOT start calling the new `PreCommitSectorBatch2` and `ProveReplicaUpdates2` methods. It isn't necessary to do so as part of this PR, and I would propose delaying calling these methods until after the nv17 upgrade.
  - See https://github.com/filecoin-project/lotus/issues/9250#issuecomment-1242766564.
- Moves things around in the state interfaces to allow for better versioning of the go-state-types objects. This is expected to be a one-time refactor, not needed every time.
## Additional Info

- [x] Needs https://github.com/filecoin-project/go-state-types/pull/68
- [x] Needs go-state-types tag
- [x] Needs https://github.com/filecoin-project/ref-fvm/pull/861 (and maybe an FVM pre-release)
- [ ] Needs https://github.com/filecoin-project/filecoin-ffi/pull/320 AFTER https://github.com/filecoin-project/ref-fvm/pull/861

## Checklist

Before you mark the PR ready for review, please make sure that:
- [ ] All commits have a clear commit message.
- [x] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [ ] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] CI is green

### CODE REVIEW

File: chain/actors/builtin/builtin.go [hunk 1]
```diff
@@ -3,24 +3,31 @@ package builtin
...
 	builtin7 "github.com/filecoin-project/specs-actors/v7/actors/builtin"
-	builtin8 "github.com/filecoin-project/specs-actors/v8/actors/builtin"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/builtin"
+	"github.com/filecoin-project/go-state-types/proof"
 
 	"github.com/filecoin-project/lotus/chain/actors"
+
+	miner9 "github.com/filecoin-project/go-state-types/builtin/v8/miner"
```
Comment (geoff-vball) [inline chain/actors/builtin/builtin.go:29]: Name/import Version mismatch

File: chain/actors/builtin/builtin.go [hunk 2]
```diff
@@ -3,24 +3,31 @@ package builtin
...
+	smoothingtypes "github.com/filecoin-project/go-state-types/builtin/v8/util/smoothing"
```
Comment (geoff-vball) [inline chain/actors/builtin/builtin.go:30]: We should probably peg this to `latestVersion`

File: chain/actors/builtin/market/v9.go [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-bitfield"
+	rlepluslazy "github.com/filecoin-project/go-bitfield/rle"
+
+	"github.com/filecoin-project/lotus/chain/actors/adt"
+	"github.com/filecoin-project/lotus/chain/types"
+
+	markettypes "github.com/filecoin-project/go-state-types/builtin/v8/market"
+	adt9 "github.com/filecoin-project/go-state-types/builtin/v8/util/adt"
```
Comment (geoff-vball) [inline chain/actors/builtin/market/v9.go:19]: version mismatch

File: chain/actors/builtin/multisig/message9.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+import (
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+
+	actorstypes "github.com/filecoin-project/go-state-types/actors"
+	builtin9 "github.com/filecoin-project/go-state-types/builtin"
+	multisig9 "github.com/filecoin-project/go-state-types/builtin/v8/multisig"
```
Comment (geoff-vball) [inline chain/actors/builtin/multisig/message9.go:11]: version mismatch

File: chain/actors/builtin/power/v9.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/ipfs/go-cid"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	"github.com/filecoin-project/lotus/chain/actors/adt"
+	"github.com/filecoin-project/lotus/chain/actors/builtin"
+
+	builtin9 "github.com/filecoin-project/go-state-types/builtin"
+	adt9 "github.com/filecoin-project/go-state-types/builtin/v8/util/adt"
```
Comment (geoff-vball) [inline chain/actors/builtin/power/v9.go:15]: version mismatch

File: chain/actors/builtin/registry.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+		for key, codeID := range codeIDs {
+			switch key {
+			case actors.AccountKey:
+				registry = append(registry, RegistryEntry{
+					code:    codeID,
+					methods: account9.Methods,
+					state:   new(account9.State),
+				})
+			}
+		}
```
Comment (geoff-vball) [inline chain/actors/builtin/registry.go:70]: I imagine these need to be extended to include the other actors?

File: chain/actors/builtin/multisig/message.go.template [hunk 7]
```diff
@@ -88,7 +90,7 @@ func (m message{{.v}}) Create(
 	return &types.Message{
 		To:     init_.Address,
 		From:   m.from,
-		Method: builtin{{.v}}.MethodsInit.Exec,
+		Method: builtin{{.latestVersion}}.MethodsInit.Exec,
```
[the remark this answers is not in the document]
Comment (arajasek) [reply chain/actors/builtin/multisig/message.go.template:93]: It's correct, but not great -- I'm gonna rewrite it to just always use the method number list in go-state-types, regardless of version. Thanks for the spot!

File: chain/actors/builtin/registry.go (hunk 6 above)
Comment (arajasek) [reply chain/actors/builtin/registry.go]: Yes, done now.

File: chain/actors/agen/main.go [hunk 8]
```diff
@@ -216,6 +223,36 @@ func generateBuiltin(builtinPath string) error {
 	return nil
 }
 
+func generateRegistry(registryPath string) error {
+
+	bf, err := ioutil.ReadFile(registryPath + ".template")
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil // skip
+		}
```
[the remark this answers is not in the document]
Comment (arajasek) [reply chain/actors/agen/main.go:232]: This is checking that the registry template file exists. I'm not sure what you mean?

[the remark this answers is not in the document]
Comment (arajasek) [reply chain/actors/agen/main.go:232]: No, this is checking that the _template_ file exists, not the registry itself.

[the remark this answers is not in the document]
Comment (geoff-vball) [reply chain/actors/agen/main.go:232]: nvm, I think I'm dumb and it's doing the opposite

**Review by geoff-vball — APPROVED** [review]

**Pull request merged**